src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,997 @@
+    means that j is the smallest value such that lo[j] does not satisfy bits. We
+    call this the first violation. i then can be computed as the largest value
+    <= j such that:
+
+    zeros[i] == lo[i] == 0
+  */
+
+  // The algorithm depends on whether the first violation violates zeros or
+  // ones, if it violates zeros, we have the bit being 1 in zero_violation and
+  // 0 in one_violation. Since all higher bits are 0 in zero_violation and
+  // one_violation, we have zero_violation > one_violation. Similarly, if the
+  // first violation violates ones, we have zero_violation < one_violation.
+  if (zero_violation < one_violation) {
+    // This means that the first bit that does not satisfy the bit requirement
+    // is a 0 that should be a 1. Obviously, since the bit at that position in
+    // ones is 1, the same bit in zeros is 0. Which means this is the value of
+    // i we are looking for.
+    //
+    // E.g:      1 2 3 4 5 6 7 8
+    //      lo = 1 1 0 0 0 1 1 0
+    //   zeros = 0 0 1 0 0 1 0 0
+    //    ones = 0 1 0 1 0 0 1 0
+    //   1-vio = 0 0 0 1 0 0 0 0
+    //   0-vio = 0 0 0 0 0 1 0 0
+    // Since the result must have the 4th bit set, it must be at least:
+    //           1 1 0 1 0 0 0 0
+    // This value must satisfy zeros, because all bits before the 4th bit have
+    // already satisfied zeros, and all bits after the 4th bit are all 0 now.
+    // Just OR this value with ones to obtain the final result.
+
+    // first_violation is the position of the violation counting from the
+    // highest bit down (0-based), since i == 4, first_violation == 3
+    juint first_violation = count_leading_zeros(one_violation);
+    //           1 0 0 0 0 0 0 0
+    constexpr U highest_bit = (std::numeric_limits<U>::max() >> 1) + U(1);
+    // This is the bit at which we want to change the bit 0 in lo to a 1, and
+    // all bits after to zero. This is similar to an operation that aligns lo
+    // up to this modulo
+    //           0 0 0 1 0 0 0 0
+    U alignment = highest_bit >> first_violation;
+    // This is the first value which have the violated bit being 1, which means
+    // that the result should not be smaller than this. This is a standard
+    // operation to align a value up to a certain power of 2.
+    // Since alignment is a power of 2, -alignment is a value having all the
+    // bits being 1 upto the location of the bit in alignment (in the example,
+    // -alignment = 11110000). As a result, lo & -alignment set all bits after
+    // the bit in alignment to 0, which is equivalent to rounding lo down to a
+    // multiple of alignment. Since lo is not divisible by alignment, to round
+    // lo up to a multiple of alignment, we add alignment to the rounded down
+    // value.
+    //           1 1 0 1 0 0 0 0
+    U new_lo = (lo & -alignment) + alignment;
+    // Our current new_lo satisfies zeros, just OR it with ones to obtain the
+    // correct result
+    //           1 1 0 1 0 0 1 0
```

Comment:
```suggestion
    // Our current new_lo satisfies zeros, just OR it with ones to obtain the
    // correct result
    //           1 1 0 1 0 0 1 0
    // Since the bits at i and before are not changed, we now have:
    // - new_lo[x] = lo[x], for 0 <= x < i        (2.5)
    // - new_lo[i] = 1                            (2.6)
    // - new_lo[x] = ones[x], for x > i           (2.7)
```